docs/site/exposing-graphql-apis.md

```diff
@@ -0,0 +1,131 @@
+---
+lang: en
+title: 'Exposing GraphQL APIs'
+keywords: LoopBack 4.0, LoopBack 4
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/exposing-graphql-apis.html
+---
+
+## Overview
+
+The [OASGraph module](https://www.npmjs.com/package/oasgraph) creates a GraphQL
+wrapper for existing REST APIs which are described by the OpenAPI specification.
+This tutorial shows how to expose GraphQL APIs in an existing LoopBack
+application.
+
+### Prerequisite
+
+Make sure you have a running LoopBack 4 application. In this tutorial, we'll use
+the `todo` example. You can create this application by running the command
+below:
+
+```sh
+lb4 example todo
+```
+
+### Customize the Server URL of the REST Server
+
+OASGraph expects the provided OpenAPI specification to have the `servers` url to
+be the server hosting the REST endpoints. We're going to modify the constructor
+of `TodoListApplication` in `src/application.ts` to include this information.
+
+```ts
+constructor(options: ApplicationConfig = {}) {
+    // Add this snippet at the beginning of the constructor
+    options = Object.assign(
+      {},
+      {
+        rest: {
+          openApiSpec: {
+            servers: [{url: 'http://localhost:3000'}],
+          },
+        },
+      },
+    );
+    // end of the code snippet
+    super(options);
+    ...
+}
+```
+
+### Install OASGraph and Required Dependencies
+
+From your LoopBack application, run the following command to install OASGraph
+and the required dependencies:
+
+```sh
+npm i --save oasgraph express-graphql
+```
+
+### Start the GraphQL Server
+
+Make sure your LoopBack application is running by going to
+`http://localhost:3000/openapi.json`. If not, you can start it by running the
+`npm start` command.
+
+Now we will use the oasgraph CLI to set up a GraphQL HTTP Server backed by
+express on port 3001. Specifying the OpenAPI spec generated by the
+todo-application as the parameter, start up the server by running the following
+command:
+
+```sh
+node_modules/.bin/oasgraph http://localhost:3000/openapi.json
+```
+
+That’s it! You’re now ready to try out some tests and requests in the browser at
+http://localhost:3001/graphql.
```

Comment:
While this is okay for building prototypes and exploring GraphQL API, I find it cumbersome to have the GraphQL endpoint running on a different port than the main REST API. It complicates configuration of gateways, load balancers and other DevOps infrastructure. Can we find a way how to server GraphQL at the same port as the REST API please? I think this work may be best left out of scope of this blog post, so let's create a new GitHub issue to keep track of this work and explain in this blog post that we are looking into ways how to server both REST and GraphQL at the same port & link to the new GitHub issue.

Thoughts?

Reply:
>  I find it cumbersome to have the GraphQL endpoint running on a different port than the main REST API. It 

Correct, the main idea of **GraphQL** library is to be used by the application. I opened #1900 for a starting point, since first we need to be able to generate the server property adequately. However, the next step is to probably :

1.  Create a CLI option `--enable-graphql` ?
2. Used the generated **OpenAPI** _schema_ by the library _(already installed by --enable-graphql in our application)_  and get an internal representation of the GraphQL Schema
3.  Use that generated GraphQL Schema to create an internal end-point (ie: /lb4graphql)  by the internal express and use the `express-graphql`  to consume that schema

That way, it will be in the same port number and u can consume graphql api from /lb4graphql .

Reply:
Thank you for opening #1900.

As I see it, the first step is to find out how to integrate `oasgraph` into LB4 application - what kind of HTTP request handler API is the module exposing? How can we plug it into an LB4 application? Do we need any any new extension points to enable this integration?

IMO, the `oasgraph` integration should live in a new extension package independent of `@loopback/core`. This package should be probably living in loopback-next alongside other LB4 modules.

I guess this is mostly what you have described in your steps 2 & 3.

Adding a CLI option like `--enable-graphql` should be the last step, the icing on a cake so to speak.

Reply:
> IMO, the oasgraph integration should live in a new extension package independent of @loopback/core 

I don't see the integration too complicated.  It only consumes the **openAPISpec** and transforms it to a **graphQLSchema** (of course more complex validations sanitations etc.), with that all we need is to use `express-graphql` as a middleware and probably reuse the same express used to expose the `openapi.json` file. 

So, in options you can disable it by setting  {  rest: {... }  ,  **oasgraph** : false }   and if that's true then we register the middleware. The middleware `express-graphql` only receives the generated graphQL Schema and take charge of everything else.

Reply:
Does it make sense to have it in a separate package?, it takes two packages and a couple of lines to implement it.

Reply:
> That way, it will be in the same port number and u can consume graphql api from /lb4graphql 

+1.  Perhaps simply /graphql?

Reply:
> Does it make sense to have it in a separate package?, it takes two packages and a couple of lines to implement it.

If we keep graphql built into our `@loopback/rest` package, then we end up with a tight coupling to a specific semver-major version of `oasgraph `. Whenever `oasgraph` makes breaking changes, we need to release a new semver-major vresion of `@loopback/rest`. 

We have been bitten by this kind of tight coupling many times in LB 3.x. I don't remember a single time when we would wish an independent extension was built into the framework. There were many times we wished a built-in feature was packaged as a standalone extension from the beginning: offline sync, authentication & authorization layer, etc.

Also by forcing all `@loopback/rest` consumers to depend on `oasgraph`, we make it harder for people not using GraphQL to avoid possible security vulnerabilities in `oasgraph` and its dependencies.

Last but not least, while `oasgraph` may be a good solution for prototyping GraphQL support, I think it's not suitable for production deployment on sites with a decent amount of traffic, because it cannot optimize database queries to fetch data from multiple related models in a single SQL query. I'd like us to look into ways how to allow LB4 app developers to leverage full power of database access and relations in the GraphQL API, see https://github.com/strongloop/loopback-next/issues/656 and https://github.com/strongloop/loopback4-extension-graphql. If we bundle `oasgraph` as the default GraphQL implementation, we are nudging app developers to use a tool that may shoot them in their foot in a near future of their project.

Reply:
Let's move this discussion to https://github.com/strongloop/loopback-next/issues/1905 please.